Run `python3 dev/update_changelog.py --prev-version 2.4...` (#9061)
mlflow/mlflow: CHANGELOG.md

Reviewer: PRs that need to be included once merged:

- https://github.com/mlflow/mlflow/pull/9053
- https://github.com/mlflow/mlflow/pull/8999

# CHANGELOG

## 2.5.0 (2023-07-13)

MLflow 2.5.0 includes several major features and improvements

> 

Reviewer: TODO: Add a section for mlflow auth

Reviewer: Ref: https://github.com/mlflow/mlflow/releases/tag/v1.27.0 where we released MLflow pipelines.

Reviewer 2: it looks great! Well done :)

Features:

- [] MLflow Gateway (#8694, @harupy)
- [Models] Add Support to the LangChain flavor for chains that contain unserializable components (#8736, @liangz1)
- [Scoring] Infer spark udf return type from model output schema (#8934, @WeichenXu123)

> - [] Add deprecation warning in gluon doc (#8971, @harupy)

Reviewer: ```suggestion
- [Models] Add deprecation warning in gluon doc (#8971, @harupy)
```

Reviewer: A separate section might make sense for this.

> - [] Add deprecation warning in gluon doc (#8971, @harupy)

Reviewer: I don't see the deprecate-gluon PR. I think it's hidden in Small bug fixes

Reviewer: Found it: https://github.com/mlflow/mlflow/pull/8968

> - [] Add deprecation warning in gluon doc (#8971, @harupy)

Reviewer: ```suggestion
- [] Deprecate the `gluon` model flavor (#8968, @harupy)
```

> - [Models] Coerce input examples into PyFunc compatible format to use for signature inference (#8860, @jerrylian-db)
> - [Models] Implement automatic signature inference for other flavors (#8782, @jerrylian-db)
> - [Models] Improve transformers signature inference implementation (#8795, @jerrylian-db)
> - [Models] Infer signature from input example when logging or saving models (#8725, @jerrylian-db)

Reviewer: ```suggestion
- [Models] Add support for automated signature inference (#8860, #8782 #8795, #8725, @jerrylian-db)
```

Bug fixes:

- [Scoring] Support nullable array type values as spark_udf return values (#9014, @WeichenXu123)
- [Models] Revert cache deletion of system modules when adding custom model code to the system path (#8722, @trungn1)
- [Models] add micro version to mlflow version pinning (#8687, @C-K-Loan)

> - [Artifacts] gc delete invalid artifact url (#8498, @PenHsuanWang)

Reviewer 2: ```suggestion
- [Artifacts] Prevent manually deleted artifacts from causing artifact garbage collection to fail (#8498, @PenHsuanWang)
```

# CHANGELOG

## 2.5.0 (2023-07-13)

MLflow 2.5.0 includes several major features and improvements

Features:

> - [] MLflow Gateway (#8694, @harupy)

Reviewer: This entry definitely needs more words or deserves its own section.

Reviewer 2: It should be a top-level announcement (banner) as we have done in the past with major releases.

Reviewer 2: Perhaps not as wordy as https://github.com/mlflow/mlflow/releases/tag/v2.0.1 , but something similar that talks about the fact that this is an entire new direction that MLflow is expanding into with the AI Gateway Server :)

> - [Docs / Examples / Models] Support chains with unserializable components (#8736, @liangz1)

Reviewer 2: ```suggestion
- [Models] Add Support to the LangChain flavor for chains that contain unserializable components (#8736, @liangz1)
```

> - [Models] [Support inference params-3]Add inference params for all flavors (#8974, @serena-ruan)

Reviewer 2: ```suggestion
```

- [Scoring] Infer spark udf return type from model output schema (#8934, @WeichenXu123)

> - [Models] [Support inference params-2]Add `params` for PyFunc predict (#8973, @serena-ruan)
> - [Models] [Support inference params-1]Add ParamSchema and ParamSpec in model signature (#8972, @serena-ruan)

Reviewer: ```suggestion
```

> - [Auth]: We are excited to announce the release of authentication and authorization support for MLflow Tracking and the MLflow Model Registry, <FILL_THIS_IN>. To get started, check out the docs at https://mlflow.org/docs/latest/auth/index.html. (#9000, #8975, #8626, #8837, #8841, @gabrielfu, @harupy)

Reviewer 2: ```suggestion
- [Auth]: We are excited to announce the release of authentication and authorization support for MLflow Tracking and the MLflow Model Registry, providing integrated access control capabilities to both services. To get started, check out the docs at https://mlflow.org/docs/latest/auth/index.html. (#9000, #8975, #8626, #8837, #8841, @gabrielfu, @harupy)
```

- [] Add deprecation warning in gluon doc (#8971, @harupy)
- [Models] Coerce input examples into PyFunc compatible format to use for signature inference (#8860, @jerrylian-db)
- [Models] Implement automatic signature inference for other flavors (#8782, @jerrylian-db)
- [Models] Improve transformers signature inference implementation (#8795, @jerrylian-db)

> - [Models] Infer signature from input example when logging or saving models (#8725, @jerrylian-db)

Reviewer 2: All of these can be collapsed into a single entry about adding support for automated signature inference :)

Reviewer: makes sense

Reviewer: https://github.com/mlflow/mlflow/pull/9061/files#r1261912173

Bug fixes:

- [Scoring] Support nullable array type values as spark_udf return values (#9014, @WeichenXu123)

> - [] Remove modifications to sys.modules (#8722, @trungn1)

Reviewer 2: ```suggestion
- [Models] Revert cache deletion of system modules when adding custom model code to the system path (#8722, @trungn1)
```

- [Models] add micro version to mlflow version pinning (#8687, @C-K-Loan)
- [Artifacts] gc delete invalid artifact url (#8498, @PenHsuanWang)

Documentation updates:

- [Docs] Update .push_model_to_sagemaker docs (#8851, @pdifranc)

> - [Docs] Fix link to Azure ML documentation 📗 (#8800, @dunnkers)

Reviewer 2: ```suggestion
- [Docs] Fix invalid link for Azure ML documentation (#8800, @dunnkers)
```

- [Models] Deprecate the `gluon` model flavor. The `mlflow.gluon` module will be removed in a future release. (#8968, @harupy)

Small bug fixes and documentation updates:

> #9048, #9043, #9035, #9034, #9037, #9038, #8993, #8966, #8985, @BenWilson2; #9039, #9036, #8902, #8924, #8866, #8861, #8810, #8761, #8544, @jerrylian-db; #8903, @smurching; #9032, #9031, #9027, #9023, #9022, #9020, #9005, #8994, #8979, #8983, #8984, #8982, #8970, #8962, #8969, #8968, #8959, #8960, #8958, #8956, #8955, #8954, #8949, #8950, #8952, #8948, #8946, #8947, #8943, #8944, #8916, #8917, #8933, #8929, #8932, #8927, #8930, #8925, #8921, #8873, #8915, #8909, #8908, #8911, #8910, #8907, #8906, #8898, #8893, #8889, #8892, #8891, #8887, #8875, #8876, #8882, #8874, #8868, #8872, #8869, #8828, #8852, #8857, #8853, #8854, #8848, #8850, #8840, #8835, #8832, #8831, #8830, #8829, #8839, #8833, #8838, #8819, #8814, #8825, #8818, #8787, #8775, #8749, #8766, #8756, #8753, #8751, #8748, #8744, #8731, #8717, #8730, #8691, #8720, #8723, #8719, #8688, #8721, #8715, #8716, #8718, #8696, #8698, #8692, #8693, #8690, @harupy; #9030, @AlimurtuzaCodes; #9029, #9025, #9021, #9013, @viktoriussuwandi; #9010, @Bncer; #9011, @Pecunia201; #9007, #9003, @EdAbati; #9002, @prithvikannan; #8991, #8867, @AveshCSingh; #8951, #8896, #8888, #8849, @gabrielfu; #8913, #8885, #8871, #8870, #8788, #8772, #8771, @serena-ruan; #8879, @maciejskorski; #7752, @arunkumarkota; #8765, #8742, #8685, #8682, #8683, @dbczumar; #8791, @mhattingpete; #8739, @yunpark93

Reviewer 2: ```suggestion
#9069, #9056, #9055, #9054, #9048, #9043, #9035, #9034, #9037, #9038, #8993, #8966, #8985, @BenWilson2; #9039, #9036, #8902, #8924, #8866, #8861, #8810, #8761, #8544, @jerrylian-db; #8903, @smurching; #9075, #9074, #9071, #9063, #9062, #9032, #9031, #9027, #9023, #9022, #9020, #9005, #8994, #8979, #8983, #8984, #8982, #8970, #8962, #8969, #8968, #8959, #8960, #8958, #8956, #8955, #8954, #8949, #8950, #8952, #8948, #8946, #8947, #8943, #8944, #8916, #8917, #8933, #8929, #8932, #8927, #8930, #8925, #8921, #8873, #8915, #8909, #8908, #8911, #8910, #8907, #8906, #8898, #8893, #8889, #8892, #8891, #8887, #8875, #8876, #8882, #8874, #8868, #8872, #8869, #8828, #8852, #8857, #8853, #8854, #8848, #8850, #8840, #8835, #8832, #8831, #8830, #8829, #8839, #8833, #8838, #8819, #8814, #8825, #8818, #8787, #8775, #8749, #8766, #8756, #8753, #8751, #8748, #8744, #8731, #8717, #8730, #8691, #8720, #8723, #8719, #8688, #8721, #8715, #8716, #8718, #8696, #8698, #8692, #8693, #8690, @harupy; #9030, @AlimurtuzaCodes; #9029, #9025, #9021, #9013, @viktoriussuwandi; #9010, @Bncer; #9011, @Pecunia201; #9007, #9003, @EdAbati; #9002, @prithvikannan; #8991, #8867, @AveshCSingh; #8951, #8896, #8888, #8849, @gabrielfu; #8913, #8885, #8871, #8870, #8788, #8772, #8771, @serena-ruan; #8879, @maciejskorski; #7752, @arunkumarkota; #8765, #8742, #8685, #8682, #8683, @dbczumar; #8791, @mhattingpete; #8739, @yunpark93
```
Adding pending PRs that will merge shortly and PRs from the last ~24hrs

## 2.5.0 (2023-07-13)

MLflow 2.5.0 includes several major features and improvements:

> - [MLflow AI Gateway] We are excited to announce the release of MLflow AI Gateway, a powerful tool designed to streamline the usage and management of various large language model (LLM) providers, such as OpenAI and Anthropic, within an organization. It offers a high-level interface that simplifies the interaction with these services by providing a unified endpoint to handle specific LLM related requests.. To get started with MLflow AI Gateway, check out the docs at https://mlflow.org/docs/latest/gateway/index.html. (#8694, @harupy, @BenWilson2, @dbczumar)

Reviewer 3: ```suggestion
- [MLflow AI Gateway] We are excited to announce the release of MLflow AI Gateway, a powerful tool designed to streamline the usage and management of various large language model (LLM) providers, such as OpenAI and Anthropic, within an organization. It offers a standardized interface that simplifies the interaction with these services and delivers centralized, secure management of credentials. To get started with MLflow AI Gateway, check out the docs at https://mlflow.org/docs/latest/gateway/index.html. (#8694, @harupy, @BenWilson2, @dbczumar)
```

> - [MLflow Authentication]: We are excited to announce the release of MLflow Authentication, <FILL_THIS_IN>. To get started with MLflow Authentication, check out the docs at https://mlflow.org/docs/latest/auth/index.html. (#9000, #8975, #8626, #8837, #8841, @gabrielfu, @harupy)

Reviewer 3: ```suggestion
- [Auth]: We are excited to announce the release of authentication and authorization support for MLflow Tracking and the MLflow Model Registry, <FILL_THIS_IN>. To get started, check out the docs at https://mlflow.org/docs/latest/auth/index.html. (#9000, #8975, #8626, #8837, #8841, @gabrielfu, @harupy)
```

Features:

- [Models] Add Support to the LangChain flavor for chains that contain unserializable components (#8736, @liangz1)
- [Scoring] Infer spark udf return type from model output schema (#8934, @WeichenXu123)
- [Models] Add support for automated signature inference (#8860, #8782 #8795, #8725, @jerrylian-db)

Bug fixes:

> - [Security] Fix critical LFI attack on Windows by enhancing `path` validation (#8999, @serena-ruan)

Reviewer 3: ```suggestion
- [Security] Improve robustness to LFI attacks on Windows by enhancing path validation (#8999, @serena-ruan)
```
